src/plugins/policy/test/main.c

```diff
@@ -0,0 +1,112 @@
+/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/* include/krb5/kdcpolicy_plugin.h - KDC policy plugin interface */
+/*
+ * Copyright (C) 2017 by Red Hat, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "k5-int.h"
+#include "kdb.h"
+#include <krb5/kdcpolicy_plugin.h>
+
+static krb5_error_code
+output_from_indicator(const char **auth_indicators,
+                      krb5_deltat *tkt_end_out, krb5_deltat *tkt_renew_out,
+                      const char **status)
+{
+    if (auth_indicators == NULL || auth_indicators[0] == NULL) {
+        *status = NULL;
+        return 0;
+    }
+
+    if (strcmp(auth_indicators[0], "fast") == 0) {
+        *tkt_end_out = 3600; /* 1:00:00 */
+        *tkt_renew_out = *tkt_end_out * 2;
+        return 0;
+    } else if (strcmp(auth_indicators[0], "pkinit") == 0) {
+        *tkt_end_out = 25200; /* 7:00:00 */
+        *tkt_renew_out = *tkt_end_out * 2;
+        return 0;
+    }
+
+    *status = "LOCAL_POLICY";
+    return KRB5KDC_ERR_POLICY;
+}
+
+static krb5_error_code
+test_local_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
+              krb5_kdc_req *request, krb5_db_entry *client,
+              krb5_db_entry *server, const char **auth_indicators,
+              const char **status, krb5_deltat *tkt_end_out,
+              krb5_deltat *tkt_renew_out)
+{
+    if (request->client != NULL &&
+        strncmp(request->client->data->data, "fail",
+                request->client->data->length) == 0) {
```

Comment:
This comparison will succeed if the first client data element is the empty string, "f", "fa", or "fai".  Although that's harmless in test code, we have had CVEs resulting from this incorrect construction in the past.  We have a helper data_eq_string() in k5-int.h to make these kind of comparisons easier.